File: Source/TransportEncoding/ZMUpdateEvent.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "ZMUpdateEvent.h"
#import "Collections+ZMTSafeTypes.h"
#import "ZMTransportData.h"
#import "NSString+UUID.h"
#import <WireTransport/WireTransport-Swift.h>

@import WireUtilities;


@interface NSDictionary (HashExtension)

- (NSUInteger)zm_hash;

@end

@implementation NSDictionary (HashExtension)

- (NSUInteger)zm_hash;
{
    NSUInteger finalHash = 0;
    for (NSObject *values in self.allValues) {
        finalHash += [values hash] * 13;
    }
    return finalHash;
}

@end


@interface ZMUpdateEvent ()

@property (nonatomic) NSDictionary *payload;
@property (nonatomic) ZMUpdateEventType type;
@property (nonatomic) ZMUpdateEventSource source;
@property (nonatomic) NSUUID *uuid;
@property (nonatomic) BOOL isTransient;
@property (nonatomic) BOOL wasDecrypted;
@property (nonatomic) NSMutableArray *debugInformationArray;

@end



@implementation ZMUpdateEvent

- (instancetype)initWithUUID:(NSUUID *)uuid payload:(NSDictionary *)payload transient:(BOOL)transient decrypted:(BOOL)decrypted source:(ZMUpdateEventSource)source
{
    self = [super init];
    if(self) {
        self.uuid = uuid;
        self.source = source;
        self.payload = payload;
        self.isTransient = transient;
        self.wasDecrypted = decrypted;
        self.debugInformationArray = [NSMutableArray array];
        [self appendDebugInformation:[NSString stringWithFormat:@"hash: %lu", (unsigned long)[payload zm_hash]]];
        if(! [self parseEventType:self.payload]) {
            return nil;
        }
    }
    return self;
}

+ (NSArray *)eventsArrayWithUUID:(NSUUID *)uuid payloadArray:(NSArray *)payloadArray transient:(BOOL)transient source:(ZMUpdateEventSource)source pushStartingAt:(NSUUID *)sourceThreshold
{
    if (payloadArray == nil) {
        ZMLogError(@"Push event payload is invalid");
        return @[];
    }
    
    NSMutableArray *events = [NSMutableArray array];
    for(NSDictionary *payload in [payloadArray asDictionaries]) {
        ZMUpdateEventSource actualSource = source;

        BOOL type1 = sourceThreshold.isType1UUID && uuid.isType1UUID;
        if (type1 && nil != sourceThreshold && [sourceThreshold compareWithType1UUID:uuid] != NSOrderedDescending) {
            actualSource = ZMUpdateEventSourcePushNotification;
        }
        ZMUpdateEvent *event = [[self alloc] initWithUUID:uuid payload:payload transient:transient decrypted:NO source:actualSource];
        if (event != nil) {
            [events addObject:event];
        }
    }
    
    return events;
}

+ (NSArray *)eventsArrayFromPushChannelData:(id<ZMTransportData>)transportData
{
    return [self eventsArrayFromTransportData:transportData source:ZMUpdateEventSourceWebSocket];
}

+ (NSArray *)eventsArrayFromPushChannelData:(id<ZMTransportData>)transportData pushStartingAt:(NSUUID *)threshold
{
    return [self eventsArrayFromTransportData:transportData source:ZMUpdateEventSourceWebSocket pushStartingAt:threshold];
}

+ (NSArray *)eventsArrayFromTransportData:(id<ZMTransportData>)transportData source:(ZMUpdateEventSource)source
{
    return [self eventsArrayFromTransportData:transportData source:source pushStartingAt:nil];
}

+ (NSArray *)eventsArrayFromTransportData:(id<ZMTransportData>)transportData source:(ZMUpdateEventSource)source pushStartingAt:(NSUUID *)threshold
{
    NSDictionary *dictionary = [transportData asDictionary];
    
    NSUUID *uuid = [[dictionary stringForKey:@"id"] UUID];
    NSArray *payloadArray = [dictionary optionalArrayForKey:@"payload"];
    BOOL transient = [dictionary optionalNumberForKey:@"transient"].boolValue;

    if(payloadArray == nil) {
        return nil;
    }
    
    if(uuid == nil) {
        ZMLogError(@"Push event id missing");
        return nil;
    }
    
    return [self eventsArrayWithUUID:uuid payloadArray:payloadArray transient:transient source:source pushStartingAt:threshold];
}

+ (nullable instancetype)eventFromEventStreamPayload:(id<ZMTransportData>)payload uuid:(NSUUID *)uuid
{
    return [[self alloc] initWithUUID:uuid payload:[payload asDictionary] transient:NO decrypted:NO source:ZMUpdateEventSourceDownload];
}

+ (nullable instancetype)decryptedUpdateEventFromEventStreamPayload:(nonnull id<ZMTransportData>)payload uuid:(nullable NSUUID *)uuid transient:(BOOL)transient source:(ZMUpdateEventSource)source
{
    return [[self alloc] initWithUUID:uuid payload:[payload asDictionary] transient:transient decrypted:YES source:source];
}

- (void)appendDebugInformation:(nonnull NSString *)debugInformation;
{
    [self.debugInformationArray addObject:debugInformation];
}

- (NSDictionary *)payload
{
    return [_payload copy];
}

- (NSString *)debugInformation
{
    return [self.debugInformationArray componentsJoinedByString:@"\n"];
}

struct TypeMap {
    CFStringRef name;
    ZMUpdateEventType type;
} const TypeMapping[] = {
    { CFSTR("call.device-info"), ZMUpdateEventCallDeviceInfo },
    { CFSTR("call.flow-active"), ZMUpdateEventCallFlowActive },
    { CFSTR("call.flow-add"), ZMUpdateEventCallFlowAdd },
    { CFSTR("call.flow-delete"), ZMUpdateEventCallFlowDelete },
    { CFSTR("call.info"), ZM_ALLOW_DEPRECATED(ZMUpdateEventCallInfo) },
    { CFSTR("call.participants"), ZMUpdateEventCallParticipants },
    { CFSTR("call.remote-candidates-add"), ZMUpdateEventCallCandidatesAdd },
    { CFSTR("call.remote-candidates-update"), ZMUpdateEventCallCandidatesUpdate },
    { CFSTR("call.remote-sdp"), ZMUpdateEventCallRemoteSDP },
    { CFSTR("call.state"), ZMUpdateEventCallState },
    { CFSTR("conversation.asset-add"), ZMUpdateEventConversationAssetAdd },
    { CFSTR("conversation.connect-request"), ZMUpdateEventConversationConnectRequest },
    { CFSTR("conversation.create"), ZMUpdateEventConversationCreate },
    { CFSTR("conversation.knock"), ZMUpdateEventConversationKnock },
    { CFSTR("conversation.member-join"), ZMUpdateEventConversationMemberJoin },
    { CFSTR("conversation.member-leave"), ZMUpdateEventConversationMemberLeave },
    { CFSTR("conversation.member-update"), ZMUpdateEventConversationMemberUpdate },
    { CFSTR("conversation.message-add"), ZMUpdateEventConversationMessageAdd },
    { CFSTR("conversation.client-message-add"), ZMUpdateEventConversationClientMessageAdd },
    { CFSTR("conversation.otr-message-add"), ZMUpdateEventConversationOtrMessageAdd },
    { CFSTR("conversation.otr-asset-add"), ZMUpdateEventConversationOtrAssetAdd },
    { CFSTR("conversation.rename"), ZMUpdateEventConversationRename },
    { CFSTR("conversation.typing"), ZMUpdateEventConversationTyping },
    { CFSTR("conversation.voice-channel"), ZMUpdateEventConversationVoiceChannel },
    { CFSTR("conversation.voice-channel-activate"), ZMUpdateEventConversationVoiceChannelActivate },
    { CFSTR("conversation.voice-channel-deactivate"), ZMUpdateEventConversationVoiceChannelDeactivate },
    { CFSTR("user.connection"), ZMUpdateEventUserConnection },
    { CFSTR("user.new"), ZMUpdateEventUserNew },
    { CFSTR("user.push-remove"), ZMUpdateEventUserPushRemove },
    { CFSTR("user.update"), ZMUpdateEventUserUpdate },
    { CFSTR("user.contact-join"), ZMUpdateEventUserContactJoin },
    { CFSTR("user.client-add"), ZMUpdateEventUserClientAdd },
    { CFSTR("user.client-remove"), ZMUpdateEventUserClientRemove },
    { CFSTR("team.create"), ZMUpdateEventTeamCreate },
    { CFSTR("team.delete"), ZMUpdateEventTeamDelete },
    { CFSTR("team.update"), ZMUpdateEventTeamUpdate },
    { CFSTR("team.member-join"), ZMUpdateEventTeamMemberJoin },
    { CFSTR("team.member-leave"), ZMUpdateEventTeamMemberLeave },
    { CFSTR("team.member-update"), ZMUpdateEventTeamMemberUpdate },
    { CFSTR("team.conversation-create"), ZMUpdateEventTeamConversationCreate },
    { CFSTR("team.conversation-delete"), ZMUpdateEventTeamConversationDelete },
};

+ (ZMUpdateEventType)updateEventTypeForEventTypeString:(NSString *)string;
{
    for (size_t i = 0; i < (sizeof(TypeMapping)/sizeof(*TypeMapping)); ++i) {
        if ([string isEqualToString:(__bridge NSString *) TypeMapping[i].name]) {
            return TypeMapping[i].type;
        }
    }
    return ZMUpdateEventUnknown;
}

+ (NSString *)eventTypeStringForUpdateEventType:(ZMUpdateEventType)type;
{
    for (size_t i = 0; i < (sizeof(TypeMapping)/sizeof(*TypeMapping)); ++i) {
        if (TypeMapping[i].type == type) {
            return (__bridge NSString *) TypeMapping[i].name;
        }
    }
    return nil;
}

- (BOOL)parseEventType:(NSDictionary *)transportData
{
    NSString *type = transportData[@"type"];
    self.type = [[self class] updateEventTypeForEventTypeString:type];
    return (self.type != ZMUpdateEventUnknown);
}

- (BOOL)isFlowEvent;
{
    switch (self.type) {
        case ZMUpdateEventCallCandidatesAdd:
        case ZMUpdateEventCallCandidatesUpdate:
        case ZMUpdateEventCallFlowActive:
        case ZMUpdateEventCallFlowAdd:
        case ZMUpdateEventCallFlowDelete:
        case ZMUpdateEventCallRemoteSDP:
            return YES;
        case ZMUpdateEventUnknown:
        case ZMUpdateEventCallParticipants:
        case ZM_ALLOW_DEPRECATED(ZMUpdateEventCallInfo):
        case ZMUpdateEventCallDeviceInfo:
        case ZMUpdateEventCallState:
        case ZMUpdateEventConversationAssetAdd:
        case ZMUpdateEventConversationConnectRequest:
        case ZMUpdateEventConversationCreate:
        case ZMUpdateEventConversationKnock:
        case ZMUpdateEventConversationMemberJoin:
        case ZMUpdateEventConversationMemberLeave:
        case ZMUpdateEventConversationMemberUpdate:
        case ZMUpdateEventConversationMessageAdd:
        case ZMUpdateEventConversationClientMessageAdd:
        case ZMUpdateEventConversationOtrMessageAdd:
        case ZMUpdateEventConversationOtrAssetAdd:
        case ZMUpdateEventConversationRename:
        case ZMUpdateEventConversationTyping:
        case ZMUpdateEventConversationVoiceChannel:
        case ZMUpdateEventConversationVoiceChannelActivate:
        case ZMUpdateEventConversationVoiceChannelDeactivate:
        case ZMUpdateEventUserConnection:
        case ZMUpdateEventUserNew:
        case ZMUpdateEventUserUpdate:
        case ZMUpdateEventUserPushRemove:
        case ZMUpdateEventUserContactJoin:
        case ZMUpdateEventUserClientAdd:
        case ZMUpdateEventUserClientRemove:
        case ZMUpdateEventTeamCreate:
        case ZMUpdateEventTeamDelete:
        case ZMUpdateEventTeamUpdate:
        case ZMUpdateEventTeamMemberJoin:
        case ZMUpdateEventTeamMemberLeave:
        case ZMUpdateEventTeamMemberUpdate:
        case ZMUpdateEventTeamConversationCreate:
        case ZMUpdateEventTeamConversationDelete:
        case ZMUpdateEvent_LAST:
            return NO;
    }
}

- (BOOL)isEncrypted
{
    switch (self.type) {
        case ZMUpdateEventConversationOtrAssetAdd:
        case ZMUpdateEventConversationOtrMessageAdd:
            return YES;
            
        default:
            return NO;
    }
}

- (BOOL)isGenericMessageEvent
{
    switch (self.type) {
        case ZMUpdateEventConversationOtrMessageAdd:
        case ZMUpdateEventConversationOtrAssetAdd:
        case ZMUpdateEventConversationClientMessageAdd:
            return true;
            break;
            
        default:
            return false;
            break;
    }
}

- (BOOL)hasEncryptedAndUnencryptedVersion
{
    switch (self.type) {
        case ZMUpdateEventConversationOtrMessageAdd:
        case ZMUpdateEventConversationOtrAssetAdd:
        case ZMUpdateEventConversationMessageAdd:
        case ZMUpdateEventConversationAssetAdd:
        case ZMUpdateEventConversationKnock:
            return YES;
            break;
            
        default:
            return NO;
            break;
    }
}

- (BOOL)isEqual:(id)object;
{
    if (! [object isKindOfClass:[ZMUpdateEvent class]]) {
        return NO;
    }
    ZMUpdateEvent *other = object;
    return (((other.uuid == self.uuid) || [other.uuid isEqual:self.uuid]) &&
            (other.type == self.type) &&
            ((other.payload == self.payload) || [other.payload isEqual:self.payload]));
}

-(NSString *)description
{
    return [NSString stringWithFormat:@"<%@ %p> %@ %@ \n %@",
            self.class, self,
            self.uuid.UUIDString,
            self.payload,
            self.debugInformation];
}


@end

File: wire-ios-transport/Source/Public/ZMUpdateEvent.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import Foundation
import WireUtilities

// MARK: - ZMUpdateEventsPolicy

@objc
public enum ZMUpdateEventsPolicy: Int {
    case buffer /// store live events in a buffer, to be processed later
    case ignore /// process events received through /notifications or /conversation/.../events
    case process /// process events received through the push channel
}

// MARK: - ZMUpdateEventSource

@objc
public enum ZMUpdateEventSource: Int {
    case webSocket
    case pushNotification
    case download
}

// MARK: - ZMUpdateEventType

@objc
public enum ZMUpdateEventType: UInt, CaseIterable, Equatable {
    case unknown = 0
    case conversationAssetAdd = 1
    case conversationConnectRequest = 2
    case conversationCreate = 3
    case conversationDelete = 39
    case conversationKnock = 4
    case conversationMemberJoin = 5
    case conversationMemberLeave = 6
    case conversationMemberUpdate = 7
    case conversationMessageAdd = 8
    case conversationClientMessageAdd = 9
    case conversationOtrMessageAdd = 10
    case conversationOtrAssetAdd = 11
    case conversationRename = 12
    case conversationProtocolUpdate = 45
    case conversationTyping = 13
    case conversationCodeUpdate = 14
    case conversationAccessModeUpdate = 15
    case conversationMessageTimerUpdate = 31
    case conversationReceiptModeUpdate = 34
    case conversationMLSWelcome = 41
    case conversationMLSMessageAdd = 42
    case userConnection = 16
    case userNew = 17
    case userUpdate = 18
    case userDelete = 35
    case userPushRemove = 19
    case userLegalHoldEnable = 38
    case userLegalHoldDisable = 37
    case userLegalHoldRequest = 36
    case userContactJoin = 20
    case userClientAdd = 21
    case userClientRemove = 22
    case userPropertiesSet = 32
    case userPropertiesDelete = 33
    case teamCreate = 23
    case teamDelete = 24
    // removed: teamUpdate = 25 [WPB-4552]: The event is no longer sent, clients must fetch team metadata (e.g. name, icon) every 24h
    // removed: teamMemberJoin = 26 [WPB-4538]: no need to handle "team.member-join" in clients
    case teamMemberLeave = 27 // [WPB-4538]: "team.member-leave" is only required for backwards compatibility
    case teamConversationCreate = 28
    case teamConversationDelete = 29
    case teamMemberUpdate = 30
    case featureConfigUpdate = 40
    case federationDelete = 43
    case federationConnectionRemoved = 44

    // Current max value: conversationProtocolUpdate = 45
}

extension ZMUpdateEventType {
    var stringValue: String? {
        switch self {
        case .unknown:
            nil
        case .conversationAssetAdd:
            "conversation.asset-add"
        case .conversationConnectRequest:
            "conversation.connect-request"
        case .conversationCreate:
            "conversation.create"
        case .conversationDelete:
            "conversation.delete"
        case .conversationKnock:
            "conversation.knock"
        case .conversationMemberJoin:
            "conversation.member-join"
        case .conversationMemberLeave:
            "conversation.member-leave"
        case .conversationMemberUpdate:
            "conversation.member-update"
        case .conversationMessageAdd:
            "conversation.message-add"
        case .conversationClientMessageAdd:
            "conversation.client-message-add"
        case .conversationOtrMessageAdd:
            "conversation.otr-message-add"
        case .conversationOtrAssetAdd:
            "conversation.otr-asset-add"
        case .conversationRename:
            "conversation.rename"
        case .conversationProtocolUpdate:
            "conversation.protocol-update"
        case .conversationTyping:
            "conversation.typing"
        case .conversationCodeUpdate:
            "conversation.code-update"
        case .conversationAccessModeUpdate:
            "conversation.access-update"
        case .conversationReceiptModeUpdate:
            "conversation.receipt-mode-update"
        case .conversationMLSWelcome:
            "conversation.mls-welcome"
        case .conversationMLSMessageAdd:
            "conversation.mls-message-add"
        case .userConnection:
            "user.connection"
        case .userNew:
            "user.new"
        case .userUpdate:
            "user.update"
        case .userDelete:
            "user.delete"
        case .userPushRemove:
            "user.push-remove"
        case .userContactJoin:
            "user.contact-join"
        case .userLegalHoldEnable:
            "user.legalhold-enable"
        case .userLegalHoldDisable:
            "user.legalhold-disable"
        case .userLegalHoldRequest:
            "user.legalhold-request"
        case .userClientAdd:
            "user.client-add"
        case .userClientRemove:
            "user.client-remove"
        case .teamCreate:
            "team.create"
        case .teamDelete:
            "team.delete"
        case .teamMemberLeave:
            "team.member-leave"
        case .teamConversationCreate:
            "team.conversation-create"
        case .teamConversationDelete:
            "team.conversation-delete"
        case .teamMemberUpdate:
            "team.member-update"
        case .conversationMessageTimerUpdate:
            "conversation.message-timer-update"
        case .userPropertiesSet:
            "user.properties-set"
        case .userPropertiesDelete:
            "user.properties-delete"
        case .featureConfigUpdate:
            "feature-config.update"
        case .federationDelete:
            "federation.delete"
        case .federationConnectionRemoved:
            "federation.connectionRemoved"
        }
    }

    init(string: String) {
        let result = ZMUpdateEventType.allCases.lazy
            .compactMap { eventType -> (ZMUpdateEventType, String)? in
                guard let stringValue = eventType.stringValue else { return nil }
                return (eventType, stringValue)
            }
            .filter { _, stringValue -> Bool in
                stringValue == string
            }
            .map { eventType, _ -> ZMUpdateEventType in
                eventType
            }
            .first

        self = result ?? .unknown
    }
}

extension ZMUpdateEvent {
    @objc(updateEventTypeForEventTypeString:)
    public static func updateEventType(for string: String)
        -> ZMUpdateEventType {
        ZMUpdateEventType(string: string)
    }

    @objc(eventTypeStringForUpdateEventType:)
    public static func eventTypeString(for eventType: ZMUpdateEventType)
        -> String? {
        eventType.stringValue
    }
}

private let zmLog = ZMSLog(tag: "UpdateEvents")

// MARK: - ZMUpdateEvent

@objcMembers
open class ZMUpdateEvent: NSObject {
    // MARK: Lifecycle

    public init?(
        uuid: UUID?,
        payload: [AnyHashable: Any]?,
        transient: Bool,
        decrypted: Bool,
        source: ZMUpdateEventSource
    ) {
        guard let payload else { return nil }
        guard let payloadType = payload["type"] as? String else { return nil }

        self.uuid = uuid
        self.payload = payload
        self.isTransient = transient
        self.wasDecrypted = decrypted

        let eventType = ZMUpdateEventType(string: payloadType)
        guard eventType != .unknown else { return nil }
        self.type = eventType
        self.source = source
        self.wasDecrypted = false
    }

    /// Creates an update event
    public convenience init?(fromEventStreamPayload payload: ZMTransportData, uuid: UUID?) {
        let dictionary = payload.asDictionary()
        // Some payloads are wrapped inside "event" key (e.g. removing bot from conversation)
        // Check for this before
        let innerPayload = (dictionary?["event"] as? [AnyHashable: Any]) ?? dictionary

        self.init(uuid: uuid, payload: innerPayload, transient: false, decrypted: false, source: .download)
    }

    // MARK: Open

    open var payload: [AnyHashable: Any]
    open var type: ZMUpdateEventType
    open var source: ZMUpdateEventSource
    open var uuid: UUID?

    /// True if the event will not appear in the notification stream
    open var isTransient: Bool
    /// True if the event had encrypted payload but now it has decrypted payload
    open var wasDecrypted: Bool

    /// True if the event is encoded with ZMGenericMessage
    open var isGenericMessageEvent: Bool {
        switch type {
        case .conversationOtrMessageAdd, .conversationOtrAssetAdd, .conversationClientMessageAdd,
             .conversationMLSMessageAdd:
            true
        default:
            false
        }
    }

    /// Debug information
    open var debugInformation: String {
        debugInformationArray.joined(separator: "\n")
    }

    open class func eventsArray(fromPushChannelData transportData: ZMTransportData) -> [ZMUpdateEvent]? {
        eventsArray(from: transportData, source: .webSocket)
    }

    /// Returns an array of @c ZMUpdateEvent from the given push channel data, the source will be set to @c
    /// ZMUpdateEventSourceWebSocket, if a non-nil @c NSUUID is given for the @c pushStartingAt parameter, all
    /// events earlier or equal to this uuid will have a source of @c ZMUpdateEventSourcePushNotification
    open class func eventsArray(
        fromPushChannelData transportData: ZMTransportData,
        pushStartingAt threshold: UUID?
    ) -> [Any]? {
        eventsArray(from: transportData, source: .webSocket, pushStartingAt: threshold)
    }

    /// Creates an update event that was encrypted and it's now decrypted
    open class func decryptedUpdateEvent(
        fromEventStreamPayload payload: ZMTransportData,
        uuid: UUID?,
        transient: Bool,
        source: ZMUpdateEventSource
    ) -> ZMUpdateEvent? {
        ZMUpdateEvent(
            uuid: uuid,
            payload: payload.asDictionary(),
            transient: transient,
            decrypted: true,
            source: source
        )
    }

    @objc(eventsArrayFromTransportData:source:)
    open class func eventsArray(from transportData: ZMTransportData, source: ZMUpdateEventSource) -> [ZMUpdateEvent]? {
        eventsArray(from: transportData, source: source, pushStartingAt: nil)
    }

    open class func eventsArray(
        from transportData: ZMTransportData,
        source: ZMUpdateEventSource,
        pushStartingAt threshold: UUID?
    ) -> [ZMUpdateEvent]? {
        let dictionary = transportData.asDictionary()
        guard let uuidString = dictionary?["id"] as? String, let uuid = UUID(uuidString: uuidString) else { return nil }
        guard let payloadArray = dictionary?["payload"] as? [Any] else { return nil }
        let transient = (dictionary?["transient"] as? Bool) ?? false

        return eventsArray(
            with: uuid,
            payloadArray: payloadArray,
            transient: transient,
            source: source,
            pushStartingAt: threshold
        )
    }

    /// Adds debug information
    open func appendDebugInformation(_ debugInformation: String) {
        debugInformationArray.append(debugInformation)
    }

    // MARK: Public

    // A hash of the event content. This is used to keep track of events
    // that we have already processed.
    public var contentHash: Int64?

    @objc(eventFromEventStreamPayload:uuid:)
    public static func eventFromEventStreamPayload(_ payload: ZMTransportData, uuid: UUID?) -> ZMUpdateEvent? {
        ZMUpdateEvent(fromEventStreamPayload: payload, uuid: uuid)
    }

    // MARK: Internal

    var debugInformationArray: [String] = []

    class func eventsArray(
        with uuid: UUID,
        payloadArray: [Any]?,
        transient: Bool,
        source: ZMUpdateEventSource,
        pushStartingAt sourceThreshold: UUID?
    ) -> [ZMUpdateEvent] {
        guard let payloads = payloadArray as? [[AnyHashable: AnyHashable]] else {
            WireLogger.updateEvent.error(
                "Push event payload is invalid",
                attributes: [.eventId: uuid.transportString().redactedAndTruncated()],
                .safePublic
            )
            return []
        }

        let events = payloads.compactMap { payload -> ZMUpdateEvent? in
            var actualSource = source
            if let thresholdUUID = sourceThreshold, thresholdUUID.isType1UUID, uuid.isType1UUID,
               (thresholdUUID as NSUUID).compare(withType1UUID: uuid as NSUUID) != .orderedDescending {
                actualSource = .pushNotification
            }
            return ZMUpdateEvent(
                uuid: uuid,
                payload: payload,
                transient: transient,
                decrypted: false,
                source: actualSource
            )
        }
        return events
    }
}

extension ZMUpdateEvent {
    override open var description: String {
        let uuidDescription = uuid?.transportString() ?? "<no uuid>"
        return "<\(Swift.type(of: self))> \(uuidDescription) \(payload) \n \(debugInformation)"
    }
}

extension ZMUpdateEvent {
    override open func isEqual(_ object: Any?) -> Bool {
        guard let other = object as? ZMUpdateEvent else {
            return false
        }

        return (uuid == other.uuid)
            && (type == other.type)
            && (payload as NSDictionary).isEqual(to: other.payload)
    }
}
